### ISSUE TITLE
In some Windows environments the indentation rule wrongly warns for SCSS maps

### ISSUE THREAD

**Issue by VPenkov** [stylelint/stylelint#1708]
I'm experiencing a bug where the Indentation plugin reports incorrect expectations.

The bug only seems to affect Windows. It was observed on multiple Windows 10 environments by myself, and sometimes on Windows 7 by @cyril-lamotte.

Observed on Node v5.5.0, v5.10.1, and v6.3.1 by both of us.

**Which rule, if any, is this issue related to?**
Indentation

**What CSS is needed to reproduce this issue?**
- [Grunt demo repository](https://github.com/VPenkov/stylelint-demo) by @VPenkov. Do ``npm install`` and run ``grunt postcss``. This is a **very** minimal repository aimed at reproducing the issue exactly.
- [Gulp demo repository](https://github.com/cyril-lamotte/stylelint-bug) by @cyril-lamotte. Do ``npm install`` and run ``gulp``. Also a minimal repository but with a little clutter in the gulpfile.
- [NPM script demo repository](https://github.com/m-allanson/stylelint-indentation-issue) by @m-allanson. Do ``npm install`` and ``npm test``.

**Which version of stylelint are you using?**
Stylelint 7.0.3

**How are you running stylelint: CLI, PostCSS plugin, Node API?**
- As a PostCSS plugin on Grunt by myself
- As a gulp-stylelint plugin by @cyril-lamotte

**Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?**
Yes, it's related to SCSS maps.

**What did you expect to happen?**
No warnings to be flagged.

**What actually happened (e.g. what warnings or errors you are getting)?**
The following warnings were flagged:"
```shell
scss/_b.font.bse_icons.scss
108:5   ‼  Expected indentation of 0 spaces (indentation) [stylelint]
```

**Comment by evilebottnawi** [comment]
@VPenkov Thanks for feedback

**Comment by evilebottnawi** [comment]
@VPenkov can't reproduce on window 10, can you create repo with reproduced error and information about node, npm, os and etc

**Comment by VPenkov** [comment]
Thanks for the quick response.
Here's a demo repo: https://github.com/VPenkov/stylelint-demo

And here's a screenshot of my error:
![image](https://cloud.githubusercontent.com/assets/1693651/17135663/8e2e0470-5339-11e6-89fc-38aa54c878e5.png)

Here's my system information:
![image](https://cloud.githubusercontent.com/assets/1693651/17135698/b8107fc0-5339-11e6-8f89-abe1067a0b4d.png)

**Comment by evilebottnawi** [comment]
@VPenkov thanks, in the near future I will try to reproduce

**Comment by VPenkov** [comment]
Thanks. Let me know if I can assist with anything else.

The issue was reproduced on a second machine with similar specs (and again Windows 10 x64) but we were unable to reproduce on Ubuntu and Windows 7.

**Comment by evilebottnawi** [comment]
@VPenkov only from repo? not npm?

**Comment by jeddy3** [comment]
@VPenkov Thanks for producing the test repo.

> A third Windows 10 instance went fine. This is not a consistent bug.

They're the worst kind of bugs.

@VPenkov Can you do us a favour and try the test repo in https://github.com/stylelint/stylelint/issues/1696#issuecomment-234887104? Can you try it on the Windows 10 machine where you were about to reproduce *your* issue and on the third Windows 10 machine where you weren't. Knowing if these two issues are related will really help us get to the bottom of things.

**Comment by VPenkov** [comment]
I was able to reproduce the bug on my personal PC as well (again Win10 x64). So 3 bugs, 1 no bug.
@evilebottnawi I was referring to the repo I've created. Otherwise we use NPM of course :)

@jeddy3 you might be onto something.
![image](https://cloud.githubusercontent.com/assets/1693651/17163833/811de0b0-53cb-11e6-91e7-3540d7600dfc.png)

**Comment by jeddy3** [comment]
@VPenkov Thanks for testing the other repo! It's good to know the issues are related.

I think the next step is to update the node version on your personal PC to `6` and try again. 

> Unfortunately it's gonna take a while to try it out on the machine where I was unable to reproduce it.

Not to worry, I think the ruling out the node version is the next priority.

**Comment by cyril-lamotte** [comment]
I updated to v6.3.1 the other day, it didn't solve the problem.
Note : I'm on "Windows 7".

**Comment by evilebottnawi** [comment]
@cyril-lamotte now i try reproduce, do not worry we will eliminate the problem :smile:

**Comment by VPenkov** [comment]
Since @cyril-lamotte's issue got closed, I've updated the title and the first post to do a better job to provide details for the issue. I'll update again and write another comment when I test with a newer Node version.

**Comment by VPenkov** [comment]
Since @cyril-lamotte's issue got closed, I've updated the title and the first post to do a better job at providing details for the issue. I can confirm the issue with Node 6.3.1.

**Comment by jeddy3** [comment]
> I've updated the title and the first post to do a better job at providing details for the issue.

Thank you for that.

> I can confirm the issue with Node 6.3.1.

As we've ruled out this being related to a:

- particular node version.
- particular windows version.

Does anyone have any ideas what the next step should be?

**Comment by evilebottnawi** [comment]
@jeddy3 can we test on `appveyor` difference window version?

**Comment by evilebottnawi** [comment]
I am working on found problem

**Comment by jeddy3** [comment]
>  can we test on appveyor difference window version?

I'm not sure. @ntwb Might know as he's the resident CI expert.

**Comment by ntwb** [comment]
No, everything is `Windows Server 2012 R2` either x64 or x86, lots of Windows based SDK's but a single platform.

**Comment by evilebottnawi** [comment]
@ntwb bad, no CI services for testing on difference window platform?

**Comment by ntwb** [comment]
I doubt it is a matter on different platforms, I'm running some tests now here myself.

My Windows Server experience of just over 20 years should be able to help with this issue 😏

**Comment by ntwb** [comment]
Testing with @VPenkov repo on Windows 10 x64 NodeJS v5.10.1 NPM 3.8.6

Adding `test.css` and testing the config `"indentation": "tab"` works as expected:
• If the indent is 4 spaces, the warning `Expected indentation of 1 tab (indentation)` is displayed
• If the indent is 1 tab no warning is displayed

Adding `test.scss` and testing the config `"indentation": "tab"` works as expected:
• If the indent is 4 spaces, the warning `Expected indentation of 1 tab (indentation)` is displayed
• If the indent is 1 tab no warning is displayed
• If the indent is 2 spaces, the warning `Expected indentation of 1 tab (indentation)` is displayed

Adding `test.css` and testing the config `"indentation": 4` works as expected:
• If the indent is 4 spaces no warning is displayed
• If the indent is 1 tab, the warning `Expected indentation of 4 spaces (indentation)` is displayed

Adding `test.scss` and testing the config `"indentation": 4` works as expected:
• If the indent is 4 spaces no warning is displayed
• If the indent is 1 tab, the warning `Expected indentation of 4 spaces (indentation)` is displayed

Adding `test.css` and testing the config `"indentation": 2` works as expected:
• If the indent is 4 spaces, the warning `Expected indentation of 2 spaces (indentation)` is displayed
• If the indent is 1 tab, the warning `Expected indentation of 2 spaces (indentation)` is displayed
• If the indent is 2 spaces no warning is displayed

Adding `test.scss` and testing the config `"indentation": 2` works as expected:
• If the indent is 4 spaces, the warning `Expected indentation of 2 spaces (indentation)` is displayed
• If the indent is 1 tab, the warning `Expected indentation of 2 spaces (indentation)` is displayed
• If the indent is 2 spaces no warning is displayed

This is crazy... Everything is working as expected... Going to start a few more alternate tests....

**Comment by ntwb** [comment]
Running `npm test` on a full stylelint checkout now... 

• NodeJS v5.10.1 NPM 3.8.6: All tests pass: `# pass  22563`
• NodeJS v6.3.1 NPM v3.10.6: All tests pass: `# pass  22563`
• `npm cache clean`, `npm install`, `npm test`: All tests pass: `# pass  22563`

-----

@VPenkov @cyril-lamotte Can you both please delete your `node_modules` folder, then run `npm cache clean`, `npm install` and try your test again

**Comment by ntwb** [comment]
Confirming that rule config: `"indentation": 4`

Using this CSS: (Indented 4 spaces)
```css
$b__char__bse_icons: (
    back-small:'\1438' // 'ᐸ'
);
```
Results in: `stylelint: F:\tmp\stylelint1708\stylelint-demo\scss\_b.font.bse_icons.scss:2:5: Expected indentation of 0 spaces (indentation)` which is incorrect

This appears to be a SCSS syntax issue, #1726 adds the first SCSS syntax test thus far for the `indentation` rule, I'm quite sure we should add some more....

It's 10:40pm here locally, I'm done for the night, but will revisit tomorrow 😄

**Comment by cyril-lamotte** [comment]
npm test doesn't seem to work, I don't know how tests work :

```$ npm test

> styleguide@1.0.0 test C:\Users\clamotte\Downloads\stylelint-demo
> echo 'Error: no test specified'

'Error: no test specified'
```

@ntwb Good night !

**Comment by ntwb** [comment]
@cyril-lamotte Sorry, `npm test` is for the stylelint repo

You should run `Gulp` ;)

Goodnight 😄

**Comment by cyril-lamotte** [comment]
Same thing :

```
$ grunt ci
Running "postcss:dist" (postcss) task

scss/_b.font.bse_icons.scss
2:5     ‼  Expected indentation of 0 spaces (indentation) [stylelint]

Fatal error:
** postcss-reporter: warnings or errors were found **
```

**Comment by m-allanson** [comment]
I've been able to replicate the issue on a Windows 10 machine with freshly installed node, git and stylelint. 

I've made a slimmed down version of @cyril-lamotte's repo to remove the gulp dependency and a few characters from the scss: https://github.com/m-allanson/stylelint-indentation-issue. Running `npm test` on that repo will trigger stylelint warnings on my Windows 10 machine, but not on my OSX machine.

It seems the bug can be triggered by multiline scss maps or lists. 

@ntwb your #1726 causes a syntax error for me on both OSX and Windows. Should we write that up as a separate issue?

**Comment by jeddy3** [comment]
@m-allanson Can you please rename `style.scss` to `style.css` (to trigger the use of the standard PostCSS parser) and try again on your Windows 10 machine? Lets rule out `postcss-scss` next.

P.S. Remember to update your npm run-scripts command too.

**Comment by m-allanson** [comment]
@jeddy3 After [these changes](https://github.com/m-allanson/stylelint-indentation-issue/commit/ead376c2b4a810783a8b1d9038a37e367df4615d) the bug is still triggered: 

![lint](https://cloud.githubusercontent.com/assets/381801/17188036/7b479286-5433-11e6-9696-3e50878c2952.png)

**Comment by m-allanson** [comment]
I can trigger the bug using http://stylelint.io/demo from the Windows machine, but not from the OSX machine.

Windows: 
(to be edited in)

OSX:
![demo-osx](https://cloud.githubusercontent.com/assets/381801/17189338/ec3f3804-5438-11e6-84f4-dc4994f887e8.png)

**Comment by m-allanson** [comment]
I've added [a failing test](https://github.com/stylelint/stylelint/issues/1729) for this issue that should work on Windows and OSX. 

I don't have any more time to look at this today. If anyone else wants to have a look, it seems like the extra `\r` characters are throwing off the `expectedIndentLevel`. The fix _might_ be [somewhere around here](https://github.com/stylelint/stylelint/blob/master/src/rules/indentation/index.js#L185-L242).

**Comment by VPenkov** [comment]
I've added @m-allanson's repo to the first post. Seems like I can't possibly contribute more to the issue but I'll hang around and watch if I can help with any more testing.

**Comment by ntwb** [comment]
Thanks everyone for your reports, testing and patience, the fix will ship as part of sylelelint 7.1

**Comment by Slapbox** [comment]
This still seems to be an issue on Windows 7.

**Comment by hudochenkov** [comment]
@Slapbox if you experience any problem, [create a new issue](https://github.com/stylelint/stylelint/issues/new/choose).
